FRAME: inherited call weight syntax

ggwpez opened the pull request:
There was a request to create a less repetitive syntax in the case that calls use the `WeightInfo` trait to calculate their weight.  

This MR adds an attribute to the `#[pallet::call]` macro in the form of `#[pallet::call(weight($type))]`.  
`$type` will then be prepended to the call name to generate a `#[pallet::weight]` attribute for each call that does not explicitly annotate one. Alternatively a `=` token can be used like `#[pallet::call(weight = …)]`.   
It is not restricted to use `<T as Config>::WeightInfo` and could use other types as well (although not advised).

The pallets `Alliance`, `Assets`, `Balances` and `example-basic` are updated to demonstrate the effect.

## Old Syntax

With the `pallet::weight` annotation above each call.

```rust
#[pallet::call]
impl<T: Config> Pallet<T> {

	#[pallet::weight(T::WeightInfo::create())]
	pub fn create(
		...
```

## New Syntax

Intended usage together with a `WeightInfo` trait. A `Config<I>` can be used for instanced pallets.

```rust
#[pallet::call(weight(<T as Config>::WeightInfo))]
impl<T: Config> Pallet<T> {

	pub fn create(
		...
```

It is possible to override the inherited value per call:
```rust
#[pallet::call(weight = <T as Config>::WeightInfo)]
impl<T: Config> Pallet<T> {

	#[pallet::weight(T::WeightInfo::not_create())]
	pub fn create(
		...
```

... or using something else entirely (for whatever reason):
```rust
struct MyWeight;
impl MyWeight {
	fn create() -> Weight {
		Weight::zero()
	}
}

#[pallet::call(weight(crate::MyWeight))]
impl<T: Config> Pallet<T> {

	pub fn create(
		...
```

### Dev Mode

The inherited weight has precedence over the default weight (=zero).

### Follow-ups
- Migrate more pallet
- Write migration script akin to the [call-index migration script](https://github.com/ggwpez/substrate-scripts/blob/master/migrate-call-index.rs)

### TODOs
- [x] Check UI errors
- [x] Tests to assert that the resulting weight value is correct.
- [x] Update / Add docs

ggwpez committed:
First approach on pallet::call_weight

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

ggwpez committed:
Use attr on pallet::call instead

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

ggwpez committed:
Ui tests

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

ggwpez committed:
Rename to weight(prefix = ...))

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2691795

KiChjang commented:
I prefer making it simple, so something like `#[pallet::call(weight($type))]` or `#[pallet::call(weight = $type)]` is more reasonable to me. Putting a `prefix` there makes people believe that you can put more attributes other than `prefix` there, which isn't true (yet).

ggwpez committed:
Simplify to #[pallet::call(weight(T))]

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

ggwpez committed:
Add stray token error

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

ggwpez committed:
Cleanup

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

ggwpez commented:
>I prefer making it simple, so something like #[pallet::call(weight($type))] or #[pallet::call(weight = $type)] is more reasonable to me. Putting a prefix there makes people believe that you can put more attributes other than prefix there, which isn't true (yet).

Yep updated it to use `weight($type)` and `weight = $type`.  
It stays extensible since we can still use keywords like `weight(const 123)`, if we ever want to.

ggwpez committed:
Migrate remaining pallets

Using script from https://github.com/ggwpez/substrate-scripts/blob/e1b5ea5b5b4018867f3e869fce6f448b4ba9d71f/frame-code-migration/src/call_weight.rs

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

ggwpez committed:
Try to add some docs

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

ggwpez committed:
Revert "Migrate remaining pallets"

Lets do this as a follow-up, I dont want to bloat this small MR.

This reverts commit 331d4b42d72de1dacaed714d69166fa1bc9c92dd.

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

ggwpez commented on frame/support/procedural/src/pallet/parse/mod.rs:
```diff
@@ -425,7 +427,36 @@ enum PalletAttr {
 	Pallet(proc_macro2::Span),
 	Hooks(proc_macro2::Span),
-	RuntimeCall(proc_macro2::Span),
+	/// A `#[pallet::call]` with optional attribute.
```
Is there a better spot to document this?  
I feel like the only way would be to create a new macro…

ggwpez replied:
I tried it yesterday. And for whatever reason the rust lang server did not like to highlight the `pallet::call` even with a stub…  
Dont know if i did it wrong, but it is annoying.

ggwpez committed:
Renames

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

sam0x17 reviewed (approved):
This looks good to me, just a few nits / comments

ggwpez committed:
Review fixes

Co-authored-by: Sam Johnson <sam@durosoft.com>

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2710632

ggwpez committed:
Test weights

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

ggwpez committed:
Update UI tests

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

ggwpez commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into oty-pallet-call-weight

muharem commented on frame/support/procedural/src/pallet/parse/mod.rs:
```diff
@@ -402,6 +402,8 @@ impl GenericKind {
 	syn::custom_keyword!(origin);
 	syn::custom_keyword!(call);
+	syn::custom_keyword!(weight);
+	syn::custom_keyword!(prefix);
```
```suggestion
```
probably forgotten to be removed

muharem commented on frame/support/procedural/src/pallet/parse/pallet_struct.rs:
```diff
@@ -52,6 +52,7 @@ pub struct PalletStructDef {
 /// * `#[pallet::generate_store($vis trait Store)]`
 /// * `#[pallet::without_storage_info]`
 /// * `#[pallet::storage_version(STORAGE_VERSION)]`
+/// * `#[pallet::call_weight($weight_trait)]`
```
probably by mistake here?

ggwpez committed:
Update frame/support/procedural/src/pallet/parse/mod.rs

Co-authored-by: Muharem Ismailov <ismailov.m.h@gmail.com>

ggwpez committed:
Remove old code

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2737281

ggwpez commented on frame/support/procedural/src/pallet/parse/call.rs:
```diff
@@ -228,17 +246,23 @@ impl CallDef {
-					FunctionAttr::Weight(w) => w,
-					_ => unreachable!("checked during creation of the let binding"),
+				let weight = match weight_attrs.len() {
+					0 if inherited_call_weight.is_some() => CallWeightDef::Inherited,
+					0 if dev_mode => CallWeightDef::DevModeDefault,
```
The question here is whether to prioritize `dev_mode` or the `call(weight = …)` when *both* are present.  
I now opted to use the specified weight instead of `dev_mode`.  

All the match statements in the expand code are exclusive, so no prioritization is needed.

ggwpez committed:
Update docs

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

ggwpez commented on frame/support/procedural/src/pallet/parse/mod.rs:
```diff
@@ -425,7 +427,36 @@ enum PalletAttr {
+	/// Both code snippets are equivalent. The passed type will be prepended to the name of the
+	/// call - if there is no `weight` attribute set on the call. This works together with
+	/// instanced pallets by using `Config<I>`. This attribute takes precedence over the default
+	/// weight if the pallet is in dev-mode.
```
I updated the docs to explain it all a bit more in-depth.

ggwpez committed:
Merge remote-tracking branch 'origin/master' into oty-pallet-call-weight

ggwpez commented:
bot merge
